File: Stormpath ObjC Example/LoginViewController.m
//
//  LoginViewController.m
//  Stormpath iOS Example
//
//  Created by Edward Jiang on 2/24/16.
//  Copyright © 2016 Stormpath. All rights reserved.
//

#import "LoginViewController.h"
#import "ProfileViewController.h"
#import "Stormpath-Swift.h"
#import "UIViewController+Alerts.h"

@interface LoginViewController ()

@property (weak, nonatomic) IBOutlet UITextField *emailTextField;
@property (weak, nonatomic) IBOutlet UITextField *passwordTextField;

@end

@implementation LoginViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self loadAccountWithFailureAlert:NO];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)loginButtonPressed:(id)sender {
    [[SPHStormpath sharedSession] loginWithUsername:self.emailTextField.text password:self.passwordTextField.text callback:^(BOOL success, NSError * _Nullable error) {
        [self loginCompletionHandler:success error:error];
    }];
}

- (IBAction)loginWithFacebookButtonPressed:(id)sender {
    [[SPHStormpath sharedSession] loginWithProvider:SPHProviderFacebook callback:^(BOOL success, NSError * _Nullable error) {
        [self loginCompletionHandler:success error:error];
    }];
}

- (IBAction)loginWithGoogleButtonPressed:(id)sender {
    [[SPHStormpath sharedSession] loginWithProvider:SPHProviderGoogle callback:^(BOOL success, NSError * _Nullable error) {
        [self loginCompletionHandler:success error:error];
    }];
}

- (void)loginCompletionHandler:(BOOL)success error:(NSError *)error {
    if(error) {
        [self showAlertWithTitle:@"Error" message:error.localizedDescription];
        return;
    }
    [self loadAccountWithFailureAlert:YES];
}

// Helper method that loads the /me endpoint (and optionally displays a
// failure alert)
- (void)loadAccountWithFailureAlert:(BOOL)failureAlert {
    [[SPHStormpath sharedSession] meWithCallback:^(SPHAccount * _Nullable account, NSError * _Nullable error) {
        if(error) {
            if(failureAlert) {
                [self showAlertWithTitle:@"Error" message:error.localizedDescription];
            }
            return;
        }
        [self showProfileForAccount:account];
    }];
}

// Helper method that takes an account object & shows the profile page
- (void)showProfileForAccount:(SPHAccount *)account {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    ProfileViewController *profileViewController = [storyboard instantiateViewControllerWithIdentifier:@"profile"];
    profileViewController.account = account;
    [self presentViewController:profileViewController animated:NO completion:nil];
}

- (IBAction)resetButtonPressed:(id)sender {
    [[SPHStormpath sharedSession] resetPasswordWithEmail:_emailTextField.text callback:^(BOOL success, NSError * _Nullable error) {
        if(success) {
            [self showAlertWithTitle:@"Success" message:@"Password Reset Email Sent!"];
        } else {
            [self showAlertWithTitle:@"Error" message:error.localizedDescription];
        }
    }];
}


/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: Stormpath Swift Example/LoginViewController.swift
//
//  LoginViewController.swift
//  Stormpath iOS Example
//
//  Created by Edward Jiang on 2/18/16.
//  Copyright © 2016 Stormpath. All rights reserved.
//

import UIKit
import Stormpath

class LoginViewController: UIViewController {
    @IBOutlet weak var emailTextField: UITextField!
    @IBOutlet weak var passwordTextField: UITextField!

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        
        loadAccount()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    @IBAction func login(_ sender: AnyObject) {
        Stormpath.sharedSession.login(emailTextField.text!, password: passwordTextField.text!, completionHandler: loginCompletionHandler)
    }
    
    @IBAction func loginWithFacebook(_ sender: AnyObject) {
        Stormpath.sharedSession.login(socialProvider: .facebook, completionHandler: loginCompletionHandler)
    }
    
    @IBAction func loginWithGoogle(_ sender: AnyObject) {
        Stormpath.sharedSession.login(socialProvider: .google, completionHandler: loginCompletionHandler)
    }
    
    func loginCompletionHandler(_ success: Bool, error: NSError?) {
        guard success else {
            self.showAlert(withTitle: "Error", message: error?.localizedDescription ?? "")
            return
        }
        self.loadAccount(andDisplayFailure: true)
    }
    
    // Helper method that loads the /me endpoint (and optionally displays a
    // failure alert)
    fileprivate func loadAccount(andDisplayFailure failureAlert: Bool = false) {
        Stormpath.sharedSession.me({ (account, error) -> Void in
            guard let account = account , error == nil else {
                if failureAlert {
                    self.showAlert(withTitle: "Error", message: error?.localizedDescription)
                }
                return
            }
            self.showProfileForAccount(account)
        })
    }
    
    // Helper method that takes an account object & shows the profile page
    func showProfileForAccount(_ account: Account) {
        let profileViewController = UIStoryboard(name: "Main", bundle: nil).instantiateViewController(withIdentifier: "profile") as! ProfileViewController
        profileViewController.account = account
        present(profileViewController, animated: false, completion: nil)
    }
    
    @IBAction func resetPassword(_ sender: AnyObject) {
        Stormpath.sharedSession.resetPassword(emailTextField.text!) { (success, error) -> Void in
            if success {
                self.showAlert(withTitle: "Success", message: "Password Reset Email Sent!")
            } else {
                self.showAlert(withTitle: "Error", message: error?.localizedDescription)
            }
        }
    }
}

// Helper extension to display alerts easily. 
extension UIViewController {
    func showAlert(withTitle title: String, message: String?) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
        self.present(alert, animated: true, completion: nil)
    }
}
